react-dom: Upgrade to v16.0.0

styfle opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://facebook.github.io/react/docs/react-dom.html#hydrate
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

vsiao commented on types/react-dom/index.d.ts:
```diff
@@ -45,6 +45,28 @@ export function render<P>(
+    element: SFCElement<P>,
+    container: Element,
+    callback?: () => any
+): void;
```
Can we write this so that it shares the same type as `render`? Consider something like:

```ts
interface Renderer {
    <P>(element: SFCElement<P>, container: Element | null, callback?: () => any): void;
    <P, T...>(element: CElement<P, T>, ...): T;
    ...
}

export const render: Renderer;
export const hydrate: Renderer;
```

harshes53 commented on types/react-dom/react-dom-tests.ts:
```diff
@@ -14,6 +14,11 @@ describe('ReactDOM', () => {
         const rootElement = document.createElement('div');
         ReactDOM.render(React.createElement('div'), rootElement);
     });
+    
```
travis failing on a trailing space on this line

harshes53 commented:
@styfle Also, not to forget to update two new APIs under `react-dom/server` e.g: `renderToNodeStream()`
and `renderToStaticNodeStream()`

styfle commented:
@plantain-00 I added v15 folder and changed the new file to v16

@vsiao I added a shared interface for both `render()` and `hydrate()`.

@harshes53 I added APIs under `react-dom/server` for `renderToNodeStream()` and `renderToStaticNodeStream()`.

Even though I have a bit of experience with TS, I rarely have to write definition files so this is a learning experience, thanks!

styfle commented:
@plantain-00 I must have messed something up because the build is the failing.

```log
Testing react
	Running: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js 
	Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/react/test/index.ts
ERROR: 506:9  expect  TypeScript@next compile error: 
Type 'Timer' is not assignable to type 'number'.
```

I don't really know where this is coming from and I didn't change the `react` typings...I only changed the `react-dom` typings.

Any help would be appreciated :)

plantain-00 commented:
You added `/// <reference types="node" />` that will import nodejs types, that will override types of `setInterval`, by default it returns `number`, in nodejs types, it return `NodeJS.Timer`
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts#L89
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/test/index.ts#L506
That is the reason, if you have to add `/// <reference types="node" />`, you need to change the `number` to `NodeJS.Timer`.

dyst5422 commented:
`createPortal is missing`

styfle commented:
@plantain-00 I'm not sure I understand the solution you suggested.
I added `/// <reference types="node" />` in the `react-dom/server/index.d.ts` file because I needed to return `NodeJS.ReadableStream`. This file does not reference any timers (or number types). Thanks for your time, please advise how to fix this problem 😄 

@dyst5422 I don't know either but I'm open to suggestions. Maybe the `createPortal` function can be added in a later PR.

styfle commented:
@plantain-00 I think I figured out the solution. I changed the test to the following:

```
private _interval: NodeJS.Timer;
```

Now the last error I'm getting is a bit confusing:

```
Testing react-dom
	Running: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js 
	Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/react-dom/index.d.ts
ERROR: 46:11  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
```

Any idea how to solve this one? 🤓

plantain-00 commented:
According to the document link, you should change `interface Renderer {` to `export interface Renderer {`

styfle commented:
@plantain-00 Thanks, that did the trick!

RyanCavanaugh commented:
Looks good to me. @vsiao can you confirm your comments have been addressed? Otherwise I will merge tomorrow.

janv commented:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18773 removed the "node" dependency from this repo. 

This PR added them again :/
Depending on `node` breaks all kinds of typings in a browser environment for me (for example, I have a custom definition for `require` that corresponds to the `require` as provided by webpack, also polyfilling `fetch`). Isn't there a way to be able to get the typings for react-dom without depending on node?

styfle commented:
> Isn't there a way to be able to get the typings for react-dom without depending on node?

The method `ReactDOM.renderToNodeStream` obviously returns a node stream so this is no accident that the return type is `NodeJS.ReadableStream`.

It might be possible to somehow move `ReadableStream` outside of the `NodeJS` namespace and maybe it wouldn't require `node`, but that seems wrong considering that it does not exist outside of the Node.js environment.

You can see an example project where this works in the browser and server here: https://github.com/styfle/react-server-example-tsx
